lib/block/blockdownload.js

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright 2021 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+'use strict';
+
+require("core-js/stable");
+
+const EventEmitter = require('events');
+const fileUrl = require('file-url');
+const DownloadError = require('./download-error');
+const { Asset } = require('../asset/asset');
+const { TransferAsset } = require('../asset/transferasset');
+const { AssetMetadata } = require('../asset/assetmetadata');
+const { TransferController, TransferEvents } = require('../controller/transfercontroller');
+const { CreateTransferParts } = require('../functions/transferpartscreate');
+const { JoinTransferParts } = require('../functions/transferpartsjoin');
+const { CloseFiles } = require('../functions/closefiles');
+const { MapConcurrent } = require('../generator/mapconcurrent');
+const { Transfer } = require('../functions/transfer');
+const { executePipeline, Pipeline } = require('../generator/pipeline');
+const { RandomFileAccess } = require('../randomfileaccess');
+const { FilterFailedAssets } = require('../functions/filterfailedassets');
+const { GetAssetMetadata } = require('../functions/getassetmetadata');
+
+/**
+ * Generate AEM download transfer assets
+ * 
+ * @generator
+ * @param {AEMDownloadOptions} options 
+ * @yields {TransferAsset} Transfer asset
+ */
+async function* generateBlockDownloadTransfer(options) {
+    console.log(`Generating block download transfer parts`);
+    const expectedLength = options.downloadFiles && (options.downloadFiles.length || options.downloadFiles.size);
+
+    let assetCounter = 0;
+    for (const downloadFile of options.downloadFiles) {
+
+        const sourceUrl = new URL(downloadFile.fileUrl);
+        const targetPath = fileUrl(downloadFile.filePath);
+
+        const source = new Asset(sourceUrl, options.headers);
+        const target = new Asset(targetPath);
+
+        const transferAsset = new TransferAsset(source, target, {
+            acceptRanges: true,
+            metadata: new AssetMetadata(downloadFile.filePath, undefined, downloadFile.fileSize)
```

Comment:
we need to pass in contentType here if its available: https://github.com/adobe/node-httptransfer/blob/f1783fdba11a4dab738c70962e89f4e6f5f042ca/lib/asset/assetmetadata.js#L30

also we need to add a test case for when all the metadata is passed in for downloadFileConcurrently